include/ofi_mem.h

```diff
@@ -88,40 +91,47 @@ static inline void* freestack_pop_impl(void *fs, void *fs_next)
 	} *freestack = fs;
 	assert(!freestack_isempty(freestack));
 	freestack->next = *((void **)fs_next);
-	return fs_next;
+	return freestack_get_user_buf_ptr(fs_next);
 }
 
 #define DECLARE_FREESTACK(entrytype, name)			\
+struct name ## _entry {						\
+	void		*ctx;					\
```

Comment:
I would just call this 'next', since we're only using it as part of a list.

We could take this change a little further and in debug builds set next to NULL on pop, and assert that it is NULL on push.